### PR TITLE
Filter out irrelevant actions

### PR DESCRIPTION
**Pull request by neuromantic99**

### CODE REVIEW

File: pages/api/v1/answers.ts [hunk 1]
```diff
@@ -457,6 +458,67 @@ async function Angela( // Good ol' Angela
...
+  if (actions.length === 0) {
+    console.log(
+      "No actions are relevant to the user's request, using simple chat model",
+    );
+    const prompt = simpleChatPrompt(
+      userQuery,
+      reqData.user_description,
+      org,
+      language,
+    );
```
Comment (Henry-Pulver) [inline pages/api/v1/answers.ts:478]: I don't really get the reason for this

It ignores all the chat history (e.g. if this happens midway through a conversation) - surely this is bad? Also, a lot of code below here is repeated further down

Couldn't we just use this prompt format in `getMessages()` if `actions === []`? Then it'll go through basically the same flow of code with knowledge of past messages (saving code and improving the responses)?

File: pages/api/v1/answers.ts [hunk 2]
```diff
@@ -774,3 +801,123 @@ async function storeActionsAwaitingConfirmation(
...
+   * Only include actions that have either:
+   * Been used previously in the conversation
+   * or
+   * Are relevant to the user's query
+   */
+
+  const actionsUsedRedisKey = conversationId.toString() + "-actions-used";
+  const actionsUsedStr = redis
+    ? ((await redis.get(actionsUsedRedisKey)) as string) ?? ""
+    : "";
```
Comment (Henry-Pulver) [inline pages/api/v1/answers.ts:823]: Don't get the reason for using redis here, we have all the past messages when we call this function, why not just grab them from there?

File: pages/api/v1/answers.ts [hunk 3]
```diff
@@ -774,3 +801,123 @@ async function storeActionsAwaitingConfirmation(
...
+    if (
+      ["0", "1", "2", "3", "4", "5", "6", "7", "8", "9"].includes(scoreString)
+    ) {
+      return Number(scoreString) > keepThreshold;
+    }
+    return true;
+  });
+}
+
+async function streamResponseToUser(
```
Comment (Henry-Pulver) [inline pages/api/v1/answers.ts:883]: Nice function

**Review by Henry-Pulver — COMMENTED** [review]
In the reasonably small amount of testing I've done, I've seen a fair few queries fail due the required functions being filtered out. In general, I think it's too harsh in filtering out - false negatives hurt way more than false positives (including functions which aren't needed)

Another reason for this might be lack of context. E.g. from the below, you wouldn't know you need  a company id to create a contact (important to know if asked to create a contact)
```
1. search_companies: Search companies. 
2. search_contacts: Search all the contacts. 
3. update_contact: Update a contact. 
4. create_contact: Create a new contact. 
5. search_deals: Search all the sales deals. 
```

In the above case, it'd probably just make up an id

Minor: it's now slower replying to simple queries that don't require API calls now since it takes 2 OpenAI calls instead of 1. Not sure there's much that can be done about this for now

File: pages/api/v1/answers.ts (hunk 2 above)
Comment (neuromantic99) [reply pages/api/v1/answers.ts:823]: Do we? We have the "nonSystemMessages" but the system prompt is generted each time from an array of actions

File: pages/api/v1/answers.ts (hunk 1 above)
Comment (neuromantic99) [reply pages/api/v1/answers.ts:478]: yep nice

**Pull request merged**